### PR TITLE
Record calls through a proxy

### LINKED ISSUE #232: Calls made using a java.lang.reflect.Proxy are recorded
Currently, only methods that are members of a class can be recorded. With this enhancement, the agent will instrument instances of java.lang.reflect.Proxy, and methods called through the proxy will be recorded.

As an example, openejb-client creates a java.lang.reflect.Proxy that the client uses to call methods on the remote. These methods will now appear in an AppMap.

### PR DESCRIPTION
**Pull request by apotterri**
Fixes #232 .

There turned out to be two aspects to this problem. 

The first was the way the agent manages javassist `ClassPool`s in a Java application server. An application server typically uses a separate ClassLoader hierarchy for each web application it loads, to isolate them from each other. The changes in 9fe3075 fix the agent so it operates correctly in such an environment.

The second aspect was enhancing the agent so it records calls made through a `java.lang.reflect.Proxy` that implements an interface that matches the config. Those changes are in 8bce824.

In addition, there are changes that fix some more checkstyle issues, as well as converting some more tests from JUnit 4 to JUnit 5.

### CODE REVIEW

**Comment by apotterri** [comment]
Preliminary version of the agent with these change: 
[appmap-1.23.0-SNAPSHOT.jar.zip](https://github.com/getappmap/appmap-java/files/13061349/appmap-1.23.0-SNAPSHOT.jar.zip)

File: agent/src/main/java/com/appland/appmap/util/AppMapClassPool.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  public static ClassPool acquire(ClassLoader classLoader) {
+    ClassPool ret = builder != null ? builder.get() : new ClassPool();
+    ret.appendClassPath(new LoaderClassPath(classLoader));
+    threadPoolQueue.get().push(ret);
+    return ret;
+  }
+
+  public static ClassPool get() {
+    return threadPoolQueue.get().peek();
```
Comment (dividedmind) [inline agent/src/main/java/com/appland/appmap/util/AppMapClassPool.java:27]: Would it make sense to assert it's not empty (also in `release()`), or is it enough to throw NPE down the line?

File: agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      logger.trace("hookSites.size(): {}", hookSites.size());
+      if (hookSites.size() == 0) {
+        // not hooked
+        return null;
+      }
+      else {
+        logger.debug("{} hooks for method: '{},'{}' behavior: {}", hookSites::size, method::toString,
+            method::toGenericString, behavior::getLongName);
+      }
+      HookSite[] ret = new HookSite[3];
```
Comment (dividedmind) [inline agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java:118]: Should this `3` be a constant in `MethodEvent` instead, for future-proofing?

File: agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+import com.appland.appmap.util.AppMapClassPool;
+import com.appland.appmap.util.ClassUtil;
+import com.appland.appmap.util.FullyQualifiedName;
+
+import javassist.CtBehavior;
+import javassist.NotFoundException;
+
+public class Proxy {
+  private static final TaggedLogger logger = AppMapConfig.getLogger(null);
+  private static final Map<Method, HookSite[]> allHookSites = new HashMap<>();
```
Comment (dividedmind) [inline agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java:31]: Should this be an `Map<Method, EnumMap<MethodEvent, HookSite>>` instead? The code is unnecessarily opaque with all those `toIndex()` and integer indices.

Or even a specialized class holding the `HookSite` as members, seeing as their catalog is rather static, and having those three `invoke`, `return`, `exc` methods that would build the event and invoke the handler.

File: agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  private static void invokeReturnHook(HookSite[] hookSites, Object proxy, Object ret, Object[] methodArgs) {
+    HookSite hookSite = hookSites[MethodEvent.METHOD_RETURN.getIndex()];
+    logger.trace("hookSite: {}, method event: {}, ret: {}", hookSite::toString, hookSite::getMethodEvent,
+        ret::toString);
+    Event event = EventTemplateRegistry.get().buildReturnEvent(hookSite.getBehaviorOrdinal());
+    MethodReturn.handle(event, proxy, ret, methodArgs);
+  }
+
+  private static void invokeExcHook(HookSite[] hookSites, Object proxy, Throwable exc, Object[] methodArgs) {
+    HookSite hookSite = hookSites[MethodEvent.METHOD_EXCEPTION.getIndex()];
```
Comment (dividedmind) [inline agent/src/main/java/com/appland/appmap/process/hooks/Proxy.java:150]: Is it possible for a hook to be missing? This isn't checked anywhere, I wonder if it should.

**Review by dividedmind — APPROVED** [review]

File: agent/src/main/java/com/appland/appmap/util/AppMapClassPool.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+  }
+
+  public static void release() {
+    threadPoolQueue.get().pop();
+  }
+
+  public static void setBuilder(Supplier<ClassPool> builder) {
+    AppMapClassPool.builder = builder;
+  }
```
[the remark this answers is not in the document]
Comment (apotterri) [reply agent/src/main/java/com/appland/appmap/util/AppMapClassPool.java:36]: I added the check in `get`. In `release`, `pop` will throw an exception if
the Deque is empty, which seems sufficient.

File: agent/src/main/java/com/appland/appmap/util/AppMapClassPool.java (hunk 1 above)
Comment (apotterri) [reply agent/src/main/java/com/appland/appmap/util/AppMapClassPool.java:27]: I added the check in `get`. In `release`, `pop` will throw an exception if
the Deque is empty, which seems sufficient.

**Pull request merged**